stock_picking_propagate_scheduled_date/models/stock_move.py

```diff
@@ -2,32 +2,65 @@
 # @author Vincent Van Rossem <vincent.vanrossem@camptocamp.com>
 # License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
 
+from datetime import timedelta
+
 from odoo import fields, models
 
 
 class StockMove(models.Model):
     _inherit = "stock.move"
 
-    def _propagate_date(self, new_date):
-        """Propagate the date of a move to its destination."""
-        already_propagated_ids = self.env.context.get(
-            "date_propagation_ids", set()
-        ) | set(self.ids)
-        self = self.with_context(date_propagation_ids=already_propagated_ids)
+    def write(self, vals):
+        previous_dates = {
+            move.id: move.date for move in self.with_context(prefetch_fields=False)
+        }
+        res = super().write(vals)
+        if "date" in vals and not self.env.context.get(
+            "_propagate_date_no_recursion", False
+        ):
+            self._propagate_date_no_recursion(previous_dates, vals.get("date"))
+        return res
+
+    def _get_moves_in_down_chain(self, chained_dest_moves=None):
+        """Get all the destination moves in the down chain of a move recursively."""
+        self.ensure_one()
+        if chained_dest_moves is None:
+            chained_dest_moves = set()
+
+        # If the dest move is already checked,
+        # return the chained_dest_moves set to avoid infinite recursion
+        if self.id in chained_dest_moves:
+            return chained_dest_moves
+
+        # Add the current move to the chained_dest_moves set
+        chained_dest_moves.add(self.id)
+
+        # Recursively find all dest moves in the down chain
+        for move_dest in self.move_dest_ids:
+            move_dest._get_moves_in_down_chain(chained_dest_moves)
+
+        return chained_dest_moves
+
+    def _propagate_date_no_recursion(self, previous_dates, new_date):
+        """Propagate the date of a move to its destination avoiding recursion."""
+        if not self:
+            return True
         for move in self:
-            if move.date:
-                delta = move.date - fields.Datetime.to_datetime(new_date)
+            prev_date = previous_dates.get(move.id)
+            if prev_date:
+                delta = prev_date - fields.Datetime.to_datetime(new_date)
             else:
-                delta = 0
-            for move_dest in move.move_dest_ids:
-                if move_dest.state in ("done", "cancel"):
-                    continue
-                if move_dest.id in already_propagated_ids:
-                    continue
-                move_dest.date -= delta
-
-    def write(self, vals):
-        # propagate date changes in the stock move chain
-        if "date" in vals:
-            self._propagate_date(vals.get("date"))
-        return super().write(vals)
+                delta = timedelta(seconds=0)
```

Comment:
I would just `continue` in this case

Reply:
It was originally proposed by @vvrossem.